### ISSUE TITLE
`misleading_variable_name` works weirdly

### ISSUE THREAD

**Issue by CleverWild** [trailofbits/dylint#1903]
It seems that `misleading_variable_name` starts matching names from the end and only a small part of them. Is this normal behavior? If so, I suggest adding a setting to this lint that allows you to specify which side to start the comparison from.

```rust
let up: UpdateSchemePath = path.into_inner(); // negative
let query: UpdateSchemeQuery = query.into_inner(); 
    ^^^^^
// `dropshot::extractor::query` exports a type `Query`, which is not     
// the type of `query` use a name other than `query`
```

**Comment by smoelius** [comment]
> It seems that `misleading_variable_name` starts matching names from the end and only a small part of them.

It works a little differently (assuming it's working correctly).

I tried to find the function called in your example, but I couldn't.

Let me please pretend it is this one, just as an example: https://docs.rs/dropshot/latest/src/dropshot/extractor/query.rs.html#31

In that case, the lint would be reacting to:
- The return type of `into_inner` is `QueryType`.
- The module `extractor::query` exports a type called `Query`.
- A reader might see the `query` variable and think its type is `Query` rather than `QueryType`.

In your example, the variable is explicitly typed, though. Looking at the lint's source code, I do not see any references to a `LetStmt`'s [`ty` field](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.LetStmt.html#structfield.ty): https://github.com/trailofbits/dylint/blob/30657a9820a8f2c0adb11e0cc8b1c15f05f8db7d/examples/restriction/misleading_variable_name/src/lib.rs#L60-L68

The lint probably should not flag cases where the variable is explicitly typed.

---

So, for anyone reading this, I would be open to a PR to:
- not flag explicitly typed variables (woudl require a small change and test)
- ways to improve the lint's documentation: https://github.com/trailofbits/dylint/blob/master/examples/restriction/misleading_variable_name

**Comment by CleverWild** [comment]
> I tried to find the function called in your example, but I couldn't.

There was used https://docs.rs/dropshot/latest/src/dropshot/extractor/body.rs.html#44.

> The lint probably should not flag cases where the variable is explicitly typed.

"Explicitly types" was written only to reveal type name in issue. There is no in my actual code. In any case, your suggestion makes sense.
